Source/Tenancy.Configuration/TenantMapManager.cs

```diff
@@ -0,0 +1,60 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Dolittle.Lifecycle;
+using Dolittle.Serialization.Json;
+using Dolittle.Types;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+
+namespace Dolittle.Tenancy.Configuration
+{
+    /// <inheritdoc/>
+    [Singleton]
+    public class TenantMapManager : ITenantMapManager
+    {
+        readonly string _path = Path.Combine(".dolittle", "tenant-map.json");
+        const string _strategyJsonKey = "strategy";
+    
+        readonly ISerializer _serializer;
+        readonly TenantStrategy _tenantStrategy;
+        readonly string _strategyConfigurationAsString;
+
+        /// <summary>
+        /// Instantiates an instance of <see cref="TenantMapManager"/>
+        /// </summary>
+        /// <param name="serializer"></param>
+        public TenantMapManager(ISerializer serializer)
+        {
+            _serializer = serializer;
+
+            var path = Path.Combine(Directory.GetCurrentDirectory(), _path);
+            
+            if (File.Exists(path)) 
+            {
+                var fileContent = File.ReadAllText(path);
```

Comment:
You should have a dependency to IFileSystem to make this testable.

Reply:
True enough. But the method :  
`public object InstanceOfStrategy(Type strategyType)`

is more interesting to do specs for - but to be able to do that you'd need to use the `IFileSystem` to be able to mock it out

Reply:
Instead of having to mock out IFileSystem and being forced to deal with FileInfo we can instead have a dependency on a system that only provides the TenantMapConfiguration (like what I did for the TenantResourceManager)

```csharp
    /// <summary>
    /// Defines a system that can provide the <see cref="ITenantResourceManager"/> with resource configurations mapped by <see cref="TenantId"/>
    /// </summary>
    public interface ICanProvideResourceConfigurationsByTenant
    {
        /// <summary>
        /// Gets a specific configuration based on the <see cref="TenantId"/> and the <see cref="ResourceType"/> 
        /// </summary>
        /// <param name="configurationType">The <see cref="Type"/> of the configuration</param>
        /// <param name="tenantId">The <see cref="TenantId"/> that the configuration belongs under</param>
        /// <param name="resourceType">The <see cref="ResourceType"/> that the configuration belongs under</param>
        /// <returns></returns>
        object ConfigurationFor(Type configurationType, TenantId tenantId, ResourceType resourceType);
        /// <summary>
        /// Gets a specific configuration based on the <see cref="TenantId"/> and the <see cref="ResourceType"/> 
        /// </summary>
        /// <param name="tenantId">The <see cref="TenantId"/> that the configuration belongs under</param>
        /// <param name="resourceType">The <see cref="ResourceType"/> that the configuration belongs under</param>
        /// <typeparam name="T">The type of configuration object</typeparam>
        /// <returns></returns>
        T ConfigurationFor<T>(TenantId tenantId, ResourceType resourceType);
    }
```
```csharp
    public class TenantResourceManager : ITenantResourceManager
    {    
        IEnumerable<IRepresentAResourceType> _resourceDefinitions;
        ICanProvideResourceConfigurationsByTenant _resourceConfigurationByTenantProvider;

        /// <summary>
        /// Instantiates an instance of <see cref="TenantResourceManager"/>
        /// </summary>
        /// <param name="resourceDefinitions"></param>
        /// <param name="resourceConfigurationByTenantProvider"></param>
        public TenantResourceManager(IInstancesOf<IRepresentAResourceType> resourceDefinitions, ICanProvideResourceConfigurationsByTenant resourceConfigurationByTenantProvider)
        {
            _resourceDefinitions = resourceDefinitions;
            _resourceConfigurationByTenantProvider = resourceConfigurationByTenantProvider;
        }
    }
```

Reply:
You could have a `TenantStrategyLoader` for something like this

Reply:
The benefit of this is that we can then isolate the loading part and would be able to add support for more than JSON as source at a later stage. A pattern we should probably look into for everything and possibly formalize this as a general principle.

Reply:
Yes, I think that is much nicer. More explicit + easier to mock and test. Explicitness is especially good